src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertProvider.cs

```diff
@@ -0,0 +1,209 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+using System.Security.Cryptography;
+using Microsoft.Win32.SafeHandles;
+
+using CURLoption = Interop.libcurl.CURLoption;
+using CURLcode  = Interop.libcurl.CURLcode;
+using CURLINFO = Interop.libcurl.CURLINFO;
+using SslClientCertCb = Interop.libcurl.client_cert_cb;
+using SafeSslContextHandle = Interop.libssl.SafeSslContextHandle;
+
+namespace System.Net.Http
+{
+    internal partial class CurlHandler : HttpMessageHandler
+    {
+        [Conditional(VerboseDebuggingConditional)]
+        private static void IgnorePeerCert(EasyRequest easy)
+        {
+            easy.SetCurlOption(CURLoption.CURLOPT_VERIFYPEER, 0L);
+        }
+
+        /// <summary>
+        ///   Uses libssl's client certificate callback to provide the client certificate.
+        ///
+        ///   Works only for libcurl with an ssl-backend that is compatible with libssl
+        /// </summary>
+        private sealed class ClientCertProvider : IDisposable
+        {
+            private static readonly Interop.libcurl.ssl_ctx_callback s_sslCallback = CurlSslCallback;
+            private readonly GCHandle _gcHandle;
+            private SafeEvpPKeyHandle _privateKeyHandle = null;
+            private SafeX509Handle _certHandle = null;
+
+            private ClientCertProvider ()
+            {
+                _gcHandle = GCHandle.Alloc(this);
+            }
+
+            public void SetClientCertificateOptions(EasyRequest easy)
+            {
+                easy.SetCurlOption(CURLoption.CURLOPT_SSL_CTX_FUNCTION, s_sslCallback);
+                easy.SetCurlOption(CURLoption.CURLOPT_SSL_CTX_DATA, GCHandle.ToIntPtr(_gcHandle));
+                IgnorePeerCert(easy);
+                VerboseTrace("setting client certificate options");
+            }
+
+            public void Dispose()
+            {
+                _gcHandle.Free();
+                if (_privateKeyHandle != null)
+                {
+                    _privateKeyHandle.Dispose();
+                }
+
+                if (_certHandle != null)
+                {
+                    _certHandle.Dispose();
+                }
+            }
+
+            private int TlsClientCertCallback(IntPtr ssl, out IntPtr certHandle, out IntPtr privateKeyHandle)
+            {
+                using (Interop.libssl.SafeSslHandle sslHandle = new Interop.libssl.SafeSslHandle(ssl))
```

Comment:
The constructor of SafeSslHandle that am using here tells its 'parent' class that the the pointer is not 'owned'. So am safe, no?